### PR TITLE
add public-activity for tracking updates in publishable records

### PR DESCRIPTION
**Pull request by kowal**
#### Summary

* Adds public_activity page
* tracks `Company, Geography, Legislation, Litigation, Target` resources
* recognizes changes to `visibility_status`, deletes and all other "edits"
* only single updates for now (bulk updates skips callbacks)

act200

### CODE REVIEW

**Comment by tsubik** [comment]
What should we do with bulk upload and bulk actions? Just one activity (disable that callback), or bunch of activities for each update. Something to consider.

**Comment by simaob** [comment]
So on the other PR I mentioned a single activity, but I guess it will depend on how much they use it, maybe for now I'd keep with tracking all of the activities individually and if see how it looks. Otherwise we might need to tweak the display so that they can see exactly what was affected.

**Comment by simaob** [comment]
One thing to add is the link to the resource, so that it's easy to go to that page and check. Other than that, this is what I wanted =D We might want to add pagination to the page, in case this grows quite large.

**Comment by tsubik** [comment]
This page could be a standard active admin index page, right? With filters, pagination, everything. Last activities could be another section that we will add to the dashboard.

**Comment by kowal** [comment]
I'll get back to this tomorrow. 
@tsubik good point with standard active admin index page, will check that.

**Comment by kowal** [comment]
> What should we do with bulk upload and bulk actions? Just one activity (disable that callback), or bunch of activities for each update. Something to consider.

Bulk updates are currently implemented using `update_all` which skips validations & callbacks, so we won't record any activity automatically. We should get more info of what we really want here and update batch actions classes accordingly but maybe let's do it i another story.

**Comment by kowal** [comment]
@simaob @tsubik please review.

I won't have much time during upcoming week, so if you find some small things to fix, feel free to take this over! thanks ;)

**Comment by kowal** [comment]
rebase after https://github.com/Vizzuality/laws_and_pathways/pull/131 was merged

**Comment by tsubik** [comment]
CSV import will generate as many activities as there were updates, maybe to be consistent bulk update should also do the same? I'm not sure.

File: app/models/concerns/public_activity_trackable.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  end
+
+  protected
+
+  def track_updates_activity
+    yield
+
+    return unless errors.blank? && previous_changes.present?
+
+    create_activity activity_key_for_last_update, owner: owner
```
Comment (kowal) [inline app/models/concerns/public_activity_trackable.rb]: If we want to manually create activities (for bulk updates, CSV imports) we could extract from here something like `#create_activity_with_current_changes`:

```ruby
module PublicActivityTrackable
  # ..
  def create_activity_with_current_changes!
    # ..
    create_activity activity_key_for_last_update, owner: owner
  end

  protected

  def track_updates_activity
    yield
    create_activity_with_current_changes!
  end
end
```

then for mentioned cases:

```ruby
company.update_attributes()
company.create_activity_with_current_changes!
```

Comment (tsubik) [reply app/models/concerns/public_activity_trackable.rb]: I think we could just start without using `update_all` in batch actions.

Comment (tsubik) [reply app/models/concerns/public_activity_trackable.rb]: Just `update` and everything should work.

Comment (tsubik) [reply app/models/concerns/public_activity_trackable.rb]: I've done in this way. Looks like it works as it should now I think

**Pull request merged**